lib/proto/build.rs

```diff
@@ -29,3 +27,42 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
 
     Ok(())
 }
+
+fn fetch_proto_dirs(
+    proto_package_dir: &Path,
+    proto_version: usize,
+    ignore_modules: Vec<&str>,
+) -> std::io::Result<Vec<PathBuf>> {
+    let version_str = format!("v{proto_version}");
+
+    let ignore_components = ignore_modules
+        .iter()
+        .map(|module| Component::Normal(OsStr::new(module)))
+        .collect::<Vec<_>>();
+
+    let proto_paths = WalkDir::new(proto_package_dir)
+        .into_iter()
+        .filter(|entry| {
+            let entry = if let Ok(entry) = entry {
+                entry
+            } else {
+                return false;
+            };
+
+            entry.path().extension().is_some_and(|ext| ext == "proto")
+                && entry
+                    .path()
+                    .components()
+                    .any(|comp| comp == Component::Normal(OsStr::new(&version_str)))
+                && !entry
+                    .path()
+                    .components()
+                    .any(|comp| ignore_components.contains(&comp))
+        })
+        .collect::<Result<Vec<_>, _>>()?;
+
+    Ok(proto_paths
+        .iter()
+        .map(|entry| entry.path().to_path_buf())
+        .collect::<Vec<PathBuf>>())
+}
```

Comment:
Is this any cleaner?

```suggestion
fn fetch_proto_dirs(
    proto_package_dir: &Path,
    proto_version: usize,
    ignore_modules: &[&str],
) -> std::io::Result<Vec<PathBuf>> {
    let version_str = format!("v{proto_version}");

    Ok(WalkDir::new(proto_package_dir)
        .into_iter()
        // Remove all files that don't have read permission
        .filter_map(|e| e.ok())
        .filter(|entry| entry.path().extension().is_some_and(|ext| ext == "proto"))
        .filter(|entry| entry.path().components().filter_map(|comp| comp.as_os_str().to_str()).any(|comp| comp == version_str))
        .filter(|entry| !entry.path().components().filter_map(|comp| comp.as_os_str().to_str()).any(|comp| ignore_modules.contains(&comp)))
        .map(|entry| entry.path().to_path_buf()).collect())
}
```